Fix #4877: Add Playlist folder sharing

Brandon-T opened the pull request:
## Security Review
- https://github.com/brave/security/issues/1005

## Summary of Changes
- Adds ability to share playlist folders (syncs from the server)
- Adds new menu UI and gets rid of bottom toolbar
- Adds loading state for syncable playlists

<!-- Enter a ticket number for this PR, create a new one if it is not there yet. -->
This pull request fixes #4877

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:
<!-- Any useful notes explaining how best to test and verify. -->


## Screenshots:
<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue and pull request is assigned to a milestone (should happen at merge time).

Brandon-T committed:
Added AsyncImage
Playlist folder sharing added

Brandon-T committed:
Redo everything

Brandon-T committed:
Add syncing logic and deleting logic

Brandon-T committed:
Add rename action

Brandon-T committed:
Added loading state

Brandon-T committed:
Fix layout and temporary crash fix

Brandon-T committed:
Fake server

Brandon-T committed:
Added management view

Brandon-T committed:
Add partial loading of playlist syncable folder, with intermediate state.

iccub reviewed (changes requested):
Needs a rebase, new setting option and we can slowly remove debug urls as there's some staging endpoint provided already

kylehickinson reviewed (changes requested):
Runtime Review:

### Preview

- Hide the toolbar on this view since its empty
- Delete in swipe options/menu doesn't seem to do anything, should probably be hidden
- Save for Offline in swipe options/menu works but then you can't delete it at all from there, probably should be hidden
- Header is broken on high font sizes: ![Simulator Screen Shot - iPhone 13 Pro - 2022-08-12 at 10 51 18](https://user-images.githubusercontent.com/529104/184389553-f11df3eb-eeee-45dc-8b0b-c82489083bb1.png)
- The add button should be using a proper image for the + symbol, not the unicode character
- You should probably hide the "back" button when adding a shared playlist
- The placeholder & final state don't match up well (the placeholders adapt to custom fonts but the final doesn't)
- The videos aren't sorted the same way each time?
- Tapping on the video player for controls immediately hides it a second later (not sure if related?)

### Add Popover

- Title needs multiline center alignment 
- The body text I feel like shouldn't be medium?
- There's a white gap at top & bottom when you get to scroll view sizes: ![Simulator Screen Shot - iPhone 13 Pro - 2022-08-12 at 10 55 37](https://user-images.githubusercontent.com/529104/184389841-a554f8b2-5d08-4165-86f7-18b287120ca3.png)
- This screen is completely broken in landscape until you get to accessibility font sizes and it becomes a scroll view. It should just probably always be a scroll view with this much text anyways: ![Simulator Screen Shot - iPhone 13 Pro - 2022-08-12 at 10 57 37](https://user-images.githubusercontent.com/529104/184389957-1fde7700-1f72-4db3-8daa-652b83598711.png)


### Post Add

- Clicking from the main list into the shared folder starts scrolled down, hiding the header entirely: ![Simulator Screen Shot - iPhone 13 Pro - 2022-08-12 at 10 58 00](https://user-images.githubusercontent.com/529104/184390019-ef03dda6-24aa-4e6e-96aa-d4ae125a3975.png)
- Toolbar is empty still, shouldn't it have move/edit/delete/etc. If its going to be edit by menu only then the toolbar should be hidden when not editing
- Delete Playlist in the menu icon isn't red and flat image icons should be replaced with custom SF Symbols: ![Simulator Screen Shot - iPhone 13 Pro - 2022-08-12 at 11 01 19](https://user-images.githubusercontent.com/529104/184390220-208df3ba-f675-4c1c-996e-3a758c57d657.png)

- The menu button is offset to far from the trailing edge of the screen, should be the same distance as the title on the leading edge
- "Remove offline data" menu item should not show if there are no items downloaded
- The scroll view's content size is wrong, has a bunch of extra space at the bottom, scroll bar doesn't hit the bottom edge: ![Simulator Screen Shot - iPhone 13 Pro - 2022-08-12 at 11 24 48](https://user-images.githubusercontent.com/529104/184390127-665ef770-f4c5-4159-abf9-fe57f265031c.png)


### Misc

- The navigation bar title seems to change colors when you scroll down… I think its unrelated to this PR but something I noticed

Brandon-T committed:
Addressed Feedback

Brandon-T committed:
Fix iPad

Brandon-T committed:
Address some feedback, localized all strings ever needed

Brandon-T committed:
Add playlist official URL

Brandon-T committed:
Accessibility

Brandon-T committed:
Accessibility Tweaks

Brandon-T committed:
Update icons

Brandon-T committed:
Undo suggestion for preferred darkmode.
Fix UUID not being re-generated everytime the tag URL changes.

Brandon-T committed:
Make everything template and remove spacer

kylehickinson reviewed (changes requested):
### Popup is still not fixed

- Landscape mode completely busted (see screenshot above)
- Smaller height screens will also break before reaching accessibility font sizes
- Settings button still does nothing

### Menu

- Icons are a bit weird, let me get you SF Symbol versions of the trash + cloud icon: <img width="73" alt="image" src="https://user-images.githubusercontent.com/529104/185682283-2f17c3bc-24d4-4b7e-a80a-83e6bfe2c09d.png">
- "Remove Offline Data" does nothing, shouldn't this toggle between save for offline and remove offline? Currently shows remove offline always

### List View (After adding the playlist)

- Multi-select checkmark doesn't show up in edit mode like in regular list
- Delete in edit mode works but no swipe to delete when you've added the list entirely
- Swipe options & long-press menus are missing when playlist is added (on purpose?)
- When layout switches for big fonts it breaks the text in each cell (true for regular playlist as well): 
![image](https://user-images.githubusercontent.com/529104/185683282-cd035018-5efe-4a4e-bb18-04c59f8bc2ca.png)

### iPad

- iPad still does nothing when launching in landscape mode, it should display the sidebar immediately

Brandon-T committed:
Feedback

Brandon-T committed:
Size button to fit any text size.
Add + image

Brandon-T committed:
Update icons

Brandon-T committed:
Add new context menu action + allow trailing swipe and context menu on persistent shared playlist folders.

Brandon-T committed:
Address bugs where the cell isn't rendering with the correct size due to `willDisplay` vs. `cellForRow`.
Added swipe to delete and context menu to shared folder cells.

Remove willDisplayCell

Brandon-T committed:
Update management popup

Brandon-T committed:
Update playlist for ETag and detecting existing folders

Brandon-T committed:
Fix videos not syncing properly due to `mobile` url vs. `desktop url` which we get from the server.

Brandon-T committed:
Use `URLString` property wrapper.
Removed `public`
Remove Encoding

Brandon-T committed:
Update images and fix rendering bugs

Brandon-T committed:
Sync every 4 hours + Autodownload

Brandon-T committed:
Sync preferences for playlist

Brandon-T committed:
Remove delete button on saved folders.
Pluralize Playlists section title.

Brandon-T committed:
Disable moving shared playlist items.
Fix font configuration for ellipsis and menu button.

Brandon-T committed:
Fix deleting. Mark the item as deleted only after popping the controller.
Fixes a bug in CoreData where the FRCs reloading causes an item to NOT be deleted (Stackoverflow)

Brandon-T committed:
Addressed feedback

kylehickinson reviewed (approved):
🎉

Brandon-T committed:
Import Helpers

Brandon-T committed:
Fixed crash on iPhone 8, iOS 14.

kylehickinson reviewed (changes requested):
Please move the last played item changes (last 2 commits) to a separate PR

Brandon-T committed:
Rewrote folder deleting.
CoreData refuses to delete the folder and merge changes asynchronously.

Brandon-T committed:
Make sure the message token is checked even though we only inject on brave owned domains.
